File: 20_07_July/0802_42892_길찾기게임/Solution.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {

    private static class Node {
        private int x;
        private int y;
        private int value;
        private Node left;
        private Node right;

        public Node(int x, int y, int value) {
            this.x = x;
            this.y = y;
            this.value = value;
        }

        public void findPreOrder(ArrayList<Integer> routine) {
            routine.add(this.value);

            if (left != null) {
                left.findPreOrder(routine);
            }

            if (right != null) {
                right.findPreOrder(routine);
            }
        }

        public void findPostOrder(ArrayList<Integer> routine) {

            if (left != null) {
                left.findPostOrder(routine);
            }

            if (right != null) {
                right.findPostOrder(routine);
            }

            routine.add(this.value);
        }

    }

    private static class Tree {
        private Node root;
        private int size;

        public void addNode(Node node) {
            ++size;

            if (root == null) {
                root = node;
                return;
            }

            Node tmp = root;

            while (true) {

                if (node.x < tmp.x) {
                    if (tmp.left == null) {
                        tmp.left = node;
                        return;
                    } else {
                        tmp = tmp.left;
                    }
                } else {
                    if (tmp.right == null) {
                        tmp.right = node;
                        return;
                    } else {
                        tmp = tmp.right;
                    }
                }

            }

        }

        public int[] getPreOrderRoutine() {
            ArrayList<Integer> routine = new ArrayList<Integer>(size);

            if (root != null) {
                root.findPreOrder(routine);
            }

            return routine.stream().mapToInt(x -> x).toArray();
        }

        public int[] getPostOrderRoutine() {
            ArrayList<Integer> routine = new ArrayList<Integer>(size);

            if (root != null) {
                root.findPostOrder(routine);
            }

            return routine.stream().mapToInt(x -> x).toArray();
        }


    }

    private int compareNodes(Node a, Node b) {
        // y 내림차순 > x 오름차순
        return (a.y != b.y) ? -Integer.compare(a.y, b.y) : Integer.compare(a.x, b.x);
    }


    public int[][] solution(int[][] nodeinfo) {
        Tree tree = new Tree();
        Node[] nodes = new Node[nodeinfo.length];

        for (int i = 0; i < nodeinfo.length; i++) {
            nodes[i] = new Node(nodeinfo[i][0], nodeinfo[i][1], i + 1);
        }

        Arrays.sort(nodes, this::compareNodes);

        for (Node node : nodes) {
            tree.addNode(node);
        }

        return new int[][] { tree.getPreOrderRoutine(), tree.getPostOrderRoutine() };
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        int[][] reuslt = sol.solution(new int[][] {{5, 3}, {11, 5}, {13, 3}, {3, 5}, {6, 1}, {1, 3}, {8, 6}, {7, 2}, {2, 2}});
    }

}

File: 20_07_July/0802_42892_길찾기게임/Solution.py
import sys

sys.setrecursionlimit(1010)

class Node(object):

    def __init__(self, x, y, value):
        self.x = x
        self.y = y
        self.value = value
        self.left = None
        self.right = None

    # 정렬을 위한 < 연산자 오버로딩
    def __lt__(self, other):
        return self.y > other.y or (self.y == other.y and self.x < other.x)
    
    def search_pre(self, routine):
        routine.append(self.value)

        if self.left != None:
            self.left.search_pre(routine)
        
        if self.right != None:
            self.right.search_pre(routine)

    def search_post(self, routine):

        if self.left != None:
            self.left.search_post(routine)

        if self.right != None:
            self.right.search_post(routine)

        routine.append(self.value)


class Tree(object):

    def __init__(self):
        self.root = None
    
    def add_node(self, node):

        if self.root == None:
            self.root = node
            return

        tmp = self.root

        while True:

            if node.x < tmp.x:
                if tmp.left == None:
                    tmp.left = node
                    return
                else:
                    tmp = tmp.left
            else:
                if tmp.right == None:
                    tmp.right = node
                    return
                else:
                    tmp = tmp.right

    def find_preorder(self):
        routine = []

        self.root.search_pre(routine)

        return routine

    def find_postorder(self):
        routine = []

        self.root.search_post(routine)

        return routine


def solution(nodeinfo):
    sorted_node_list = sorted(map(lambda x: Node(x[1][0], x[1][1], x[0] + 1), enumerate(nodeinfo)))
    tree = Tree()

    for node_info in sorted_node_list:
        tree.add_node(node_info)

    return [tree.find_preorder(), tree.find_postorder()]

print(solution([[5,3],[11,5],[13,3],[3,5],[6,1],[1,3],[8,6],[7,2],[2,2]]))

tmp = []

for i in range(1000):
    tmp.append([i, i])

print(solution(tmp))
